numpy/core/src/multiarray/ctors.c

```diff
@@ -477,32 +471,39 @@ setArrayFromSequence(PyArrayObject *a, PyObject *s,
         return 0;
     }
 
+    /*
+     * This code is to ensure that the sequence access below will
+     * return a lower-dimensional sequence.
+     */
+
     if (dim > PyArray_NDIM(a)) {
         PyErr_Format(PyExc_ValueError,
                  "setArrayFromSequence: sequence/array dimensions mismatch.");
         goto fail;
     }
 
-    tmp = _array_from_array_like(s, /*dtype*/NULL, /*writeable*/0, /*context*/NULL);
-    if (tmp == NULL) {
+    slen = PySequence_Length(s);
+    if (slen < 0) {
         goto fail;
     }
-    else if (tmp != Py_NotImplemented) {
-        if (PyArray_CopyInto(dst, (PyArrayObject *)tmp) < 0) {
+    if (slen > 0) {
+        /* gh-13659: objects with 0 len can still have ndim > ndim(dst) */
+        PyObject *tmp = _array_from_array_like(s, /*dtype*/NULL, /*writeable*/0,
+                                               /*context*/NULL);
+        if (tmp == NULL) {
             goto fail;
         }
+        else if (tmp != Py_NotImplemented) {
+            if (PyArray_CopyInto(dst, (PyArrayObject *)tmp) < 0) {
+                goto fail;
```

Comment:
```suggestion
                Py_DECREF(tmp);
                goto fail;
```
Something is still off here with the reference counting. Aso the return 0 below needs it. Which probably means the whether or not the else clause is there did not matter much. Or we make `tmp = NULL;` one scope up to make the fail paths nicer.

Reply:
refactored

Reply:
Thanks, looks right now. Will merge as soon as tests finish (or squash merge).